cpp/src/arrow/testing/math.cc

```diff
@@ -0,0 +1,95 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/testing/math.h"
+
+#include <cmath>
+#include <limits>
+
+#include <gtest/gtest.h>
+
+#include "arrow/util/logging.h"
+
+namespace arrow {
+namespace {
+
+template <typename Float>
+constexpr Float kOneUlp;
+template <>
+constexpr float kOneUlp<float> = 5.9604645e-08f;
+template <>
+constexpr double kOneUlp<double> = 1.1102230246251565e-16;
+
+template <typename Float>
+bool WithinUlpOneWay(Float left, Float right, int n_ulp) {
+  // Ideally these would be compile-time (constexpr) checks
+  DCHECK_LT(Float(1.0) - kOneUlp<Float>, Float(1.0));
+  DCHECK_EQ(std::nextafter(Float(1.0) - kOneUlp<Float>, Float(1.0)), Float(1.0));
```

Comment:
How can these `DCHECK`s fail (suppose no one ever changes the `kOneUlp` definition)? Platform-specific issues?

Reply:
They could fail if the `kOneUlp` definitions were wrong :)

Reply:
Or, yes, potential platform- or compiler-specific issues with FP literals and FP representation?